R/predict.R

```diff
@@ -0,0 +1,512 @@
+      n = n,
+      mean_delay = mean_delay,
+      seed = seed,
+      ...
+    )
+  } else if (parameter == "Rt") {
+    predict_rt(
+      object = object,
+      horizon = horizon,
+      min_date = min_date,
+      max_date = max_date,
+      n = n,
+      mean_delay = mean_delay,
+      seed = seed,
+      gi_pmf = gi_pmf,
+      ...
+    )
+  }
+}
+
+#' Parameter-specific posterior predictions
+#'
+#' @inherit predict.RtGam
+#' @param call The calling environment to pass to error messages
+#' @name predictor
+NULL
+
+#' @rdname predictor
+# " @export
+predict_obs_cases <- function(
+    object,
+    horizon = NULL,
+    min_date = NULL,
+    max_date = NULL,
+    n = 10,
+    gi_pmf = NULL,
+    seed = 12345,
+    call = rlang::caller_env(),
+    ...) {
+  newdata <- create_newdata_dataframe(
+    object = object,
+    parameter = "obs_cases",
+    min_date = min_date,
+    max_date = max_date,
+    horizon = horizon
+  )
+
+  # Use `posterior_samples()` over `fitted_samples()` to get response
+  # w/ obs uncertainty
+  fitted <- gratia::posterior_samples(
+    object[["model"]],
+    data = newdata,
+    unconditional = TRUE,
+    n = n,
+    seed = seed,
+    ...
+  )
+  format_predicted_dataframe(parameter = "obs_cases", fitted, newdata)
+}
+
+#' @rdname predictor
+#' @export
+#' @export
+predict_obs_incidence <- function(
+    object,
+    horizon = NULL,
+    min_date = NULL,
+    max_date = NULL,
+    n = 10,
+    gi_pmf = NULL,
+    seed = 12345,
+    mean_delay,
+    call = rlang::caller_env(),
+    ...) {
+  newdata <- create_newdata_dataframe(
+    object = object,
+    parameter = "obs_incidence",
+    mean_delay = mean_delay,
+    min_date = min_date,
+    max_date = max_date,
+    horizon = horizon
+  )
+  # Use `posterior_samples()` over `fitted_samples()` to get response
+  # w/ obs uncertainty
+  fitted <- gratia::posterior_samples(
+    object[["model"]],
+    data = newdata,
+    unconditional = TRUE,
+    n = n,
+    seed = seed,
+    ...
+  )
+  format_predicted_dataframe(
+    parameter = "obs_incidence",
+    fitted, newdata
+  )
+}
+
+#' @rdname predictor
+# " @export
+predict_growth_rate <- function(
+    object,
+    horizon = NULL,
+    min_date = NULL,
+    max_date = NULL,
+    n = 10,
+    gi_pmf = NULL,
+    seed = 12345,
+    mean_delay,
+    call = rlang::caller_env(),
+    ...) {
+  delta <- compute_delta(object)
+  newdata <- create_newdata_dataframe(
+    object = object,
+    parameter = "r",
+    mean_delay = mean_delay,
+    min_date = min_date,
+    max_date = max_date,
+    horizon = horizon,
+    delta = delta
+  )
+  fitted <- gratia::fitted_samples(
+    object[["model"]],
+    data = newdata,
+    n = n,
+    seed = seed,
+    unconditional = TRUE,
+    scale = "linear_predictor",
+    ...
+  )
+  format_predicted_dataframe(
+    parameter = "r",
+    fitted = fitted,
+    newdata = newdata,
+    delta = delta
+  )
+}
+
+#' @rdname predictor
+#' @export
+predict_rt <- function(
```

Comment:
As implemented these all seem like they could be very light wrappers around an internal function that passed the string vs having so much boiler plate

Reply:
After F2f we decided to leave as is and plan to turn these into generics to dispatch on backend once future backends are added.